lua/telescope/builtin/git.lua

```diff
@@ -77,32 +78,81 @@ git.bcommits = function(opts)
 end
 
 git.branches = function(opts)
-  local output = utils.get_os_command_output({ 'git', 'branch', '--all' }, opts.cwd)
+  local format = '{'
+              ..   '"head":%(if:equals=*)%(HEAD)%(then)true%(else)false%(end)'
+              ..   ',"refname":"%(refname)"'
+              ..   ',"authorname":"%(authorname)"'
+              ..   '%(if)%(upstream)%(then)'
+              ..     ',"upstream":"%(upstream:lstrip=2)"'
+              ..   '%(else)'
+              ..     ',"upstream":""'
+              ..   '%(end)'
+              ..   ',"committerdate":"%(committerdate:format-local:%Y/%m/%d %H:%M:%S)"'
+              .. '}'
+  local output = utils.get_os_command_output({ 'git', 'for-each-ref', '--format', format }, opts.cwd)
 
   local results = {}
+  local widths = {
+    name = 0,
+    authorname = 0,
+    upstream = 0,
+    committerdate = 0,
+  }
+  local register_entry = function(entry, trim_refname_prefix)
+    entry.name = string.sub(entry.refname, string.len(trim_refname_prefix)+1)
+    for key, value in pairs(widths) do
+      widths[key] = math.max(value, vim.fn.strdisplaywidth(entry[key]))
+    end
+    if string.len(entry.upstream) > 0 then
+      widths.upstream_indicator = 2
+    end
+    table.insert(results, entry)
+  end
   for _, v in ipairs(output) do
-    if not string.match(v, 'HEAD') and v ~= '' then
-      if vim.startswith(v, '*') then
-        table.insert(results, 1, v)
-      else
-        table.insert(results, v)
-      end
+    local entry = vim.fn.json_decode(v)
+    if entry.head then
+      goto continue
```

Comment:
Can we please not use `goto`? It breaks lua 5.1 and some people have to build neovim with 5.1 
Example: https://github.com/nvim-telescope/telescope.nvim/issues/518